types/npmcli__package-json/index.d.ts

```diff
@@ -0,0 +1,21 @@
+// Type definitions for @npmcli/package-json 2.0
+// Project: https://github.com/npm/package-json
+// Definitions by: Michaël De Boey <https://github.com/MichaelDeBoey>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+import type { PackageJson } from './package-json';
+
+export type Content = PackageJson;
+
+declare class NPMCliPackageJson {
+    constructor(path: string);
+
+    static load: (path?: string) => Promise<NPMCliPackageJson>;
+    save: () => Promise<void>;
+    update: (content: Content) => void;
+
+    content: Content;
+}
+
+export type { PackageJson } from './package-json';
+export default NPMCliPackageJson;
```

Comment:
at the moment, IMO this is proper CJS module:
https://github.com/npm/package-json/blob/6007cad20f83e2dbc80d3453abba3c01a660df83/lib/index.js#L109
so that would require:
```suggestion
export = PackageJson;
```

Reply:
That's not possible according to the compiler

![](https://user-images.githubusercontent.com/6643991/167272318-d941f85e-4227-4040-af4f-50733e5fdf18.png)

Reply:
that's OK, you would have to either use a symbol to export everything (excluding types) or use a namespace. I would try this:
```ts
import type { PackageJson as PackageJsonType } from './package-json';

declare class PackageJson {
    static load: (path?: string) => Promise<PackageJson>;
    constructor(path: string);

    save: () => Promise<void>;
    update: (content: PackageJson.Content) => void;

    content: PackageJson.Content;
}

declare namespace PackageJson {
    type Content = PackageJsonType;
}

export = PackageJson;
```
(please rename class to a proper name used in the module)

Reply:
Changing the code to

```ts
import type { PackageJson as PackageJsonType } from './package-json';

declare class NPMCliPackageJson {
    constructor(path: string);

    /** @async */
    static load: (path?: string) => Promise<NPMCliPackageJson>;
    save: () => Promise<void>;
    update: (content: NPMCliPackageJson.Content) => void;

    content: NPMCliPackageJson.Content;
}

declare namespace NPMCliPackageJson {
    type Content = PackageJsonType;

    type PackageJson = PackageJsonType;
}

export = NPMCliPackageJson;
```

Would give the following error in the test file

![](https://user-images.githubusercontent.com/6643991/167317116-f6e94929-c199-49d6-bf41-5917326fa5cf.png)

Reply:
the error is correct, so please update the rest of the code, in case of imports:
```ts
import ... = require('....');
```
https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require
or:
```ts
import * as .... from '...';
```
(also supported).
The DT modules promotes interop between types definitions. When consumed by users, 'esModuleInterop' is probably already set as being default from some time in TS, so this:
```ts
import .... from '...';
```
would work out of the box.
https://github.com/DefinitelyTyped/DefinitelyTyped#esmoduleinteropallowsyntheticdefaultimports

Reply:
@peterblazejewicz That would make the test pass again, but that's not how the package is meant to be used

It's meant to be used like

```ts
import PackageJson from '@npmcli/package-json';
```

Reply:
Yes, I understand:

> When consumed by users, 'esModuleInterop' is probably already set as being default from some time in TS, so this:

IMO this module is a classic Node CJS, without default export or default export alias:
https://github.com/npm/package-json/blob/main/lib/index.js#L109
https://www.typescriptlang.org/tsconfig#esModuleInterop

Reply:
Let me rephrase. I'm asking to fix definition for purpose of DefinitelyType and TS. 
It will work perfectly OK for users downloading those types.

Reply:
OK I don't think I understand correctly then 🤔

The error I got was because I changed the code to how you suggested it should be.
So what do you think I should do exactly to make it all work?

Reply:
it's not me, this is just advised, I could write:
```ts
import type { PackageJson as PackageJsonType } from './package-json';

declare namespace PackageJson {
    type Content = PackageJsonType;
}

declare class PackageJson {
    static load: (path?: string) => Promise<PackageJson>;
    constructor(path: string);

    save: () => Promise<void>;
    update: (content: PackageJson.Content) => void;

    content: PackageJson.Content;
}

export = PackageJson;

// this export class with proper name and re-exports your types via namespace export

// test within DT repo
import PackageJson = require('@npmcli/package-json');

(async () => {
    const pkgJson = await PackageJson.load('./');

    pkgJson.update({
        dependencies: {
            a: '^1.0.0',
            b: '^1.2.3',
        },
        workspaces: ['./new-workspace'],
    });

    await pkgJson.save();
})();

// usage via default TS configs would be
// import PackageJson from '@npmcli/package-json';
// or const PackageJson = require('@npmcli/package-json');
```

Reply:
If people could still do

```ts
import PackageJson from '@npmcli/package-json';
```

Then necessary changes are made.